### ISSUE TITLE
Make adding a new RPC to .proto file non-breaking for service implementation (server-side)

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#5371]
Currently, services are implemented by implementing an autogenerated interface (e.g. IMath). When a new RPC is added to to the .proto file, regenerating the code leads to breaking a build, because a new method is added to the serverside stub interface (which is quite natural behavior as such, but can be very undesirable at times).

A remedy would be to auto-generate an   FooImplBase (naming TBD) that provides virtual methods with default implementation of all RPC returning just StatusCode.UNIMPLEMENTED.
If service implementation is inherited from the base class mentioned, adding new RPCs to the proto file does no longer break user's code.

**Comment by jtattermusch** [comment]
FYI @jskeet  @chrisdunelm @ejona86   with who I'd discuss this topic previously.

**Comment by chrisdunelm** [comment]
Are you proposing that the (eg) IMath interface is still generated alongside the new base class? Or that only the base class would be generated (so no IMath interface)?

**Comment by jskeet** [comment]
Three proposals for the client side (showing just one operation):

Option 1 - four methods per operation, no reliance on `CallInvocationDetails` in the abstract class
---

```csharp
public abstract class AbstractMathClient
{
    public virtual DivReply Div(DivArgs args, CallOptions options)
    {
        throw new NotImplementedException();
    }

    public virtual AsyncUnaryCall<DivReply> DivAsync(DivArgs args, CallOptions options)
    {
        throw new NotImplementedException();
    }

    // Note: non-virtual
    public DivReply Div(DivArgs args, Metadata headers = null, DateTime? deadline = null, CancellationToken cancellationToken = default(CancellationToken))
    {
        return Div(args, new CallOptions(headers, deadline, cancellationToken));
    }

    public AsyncUnaryCall<DivReply> DivAsync(DivArgs args, Metadata headers = null, DateTime? deadline = null, CancellationToken cancellationToken = default(CancellationToken))
    {
        return DivAsync(args, new CallOptions(headers, deadline, cancellationToken));
    }
}

public sealed class MathClient : AbstractMathClient
{
    private readonly Channel channel;

    // TODO: Header interceptors? Host? Combine them all into one concept?
    public MathClient(Channel channel)
    {
        this.channel = channel; // Plus validation :)
    }

    public override DivReply Div(DivArgs args, CallOptions options)
    {
        // CreateCall could be on channel, potentially...
        var call = SomeHelperType.CreateCall(channel, __Method_Div, options);
        return Calls.BlockingUnaryCall(call, request);
    }

    public override AsyncUnaryCall<DivReply> DivAsync(DivArgs args, CallOptions options)
    {
        var call = SomeHelperType.CreateCall(channel, __Method_Div, options);
        return Calls.AsyncUnaryCall(call, request);
    }
}
```

Option 2
---

Four *concrete* methods on the abstract class, and one virtual:

```csharp
public abstract class AbstractMathClient
{
    public DivReply Div(DivArgs args, CallOptions options)
    {
        var call = CreateDivCall(options);
        return Calls.BlockingUnaryCall(call, request);
    }

    public AsyncUnaryCall<DivReply> DivAsync(DivArgs args, CallOptions options)
    {
        var call = CreateDivCall(options);
        return Calls.AsyncUnaryCall(call, request);
    }

    public DivReply Div(DivArgs args, Metadata headers = null, DateTime? deadline = null, CancellationToken cancellationToken = default(CancellationToken))
    {
        return Div(args, new CallOptions(headers, deadline, cancellationToken));
    }

    public AsyncUnaryCall<DivReply> DivAsync(DivArgs args, Metadata headers = null, DateTime? deadline = null, CancellationToken cancellationToken = default(CancellationToken))
    {
        return DivAsync(args, new CallOptions(headers, deadline, cancellationToken));
    }

    protected virtual CallInvocationDetails<DivArgs, DivReply> CreateDivCall(CallOptions options)
    {
        throw new NotImplementedException();
    }
}

public sealed class MathClient : AbstractMathClient
{
    private readonly Channel channel;

    // TODO: Header interceptors? Host? Combine them all into one concept?
    public MathClient(Channel channel)
    {
        this.channel = channel; // Plus validation :)
    }

    protected override CallInvocationDetails<DivArgs, DivReply> CreateDivCall(CallOptions options)
    {
        return SomeHelperType.CreateCall(channel, __Method_Div, options);
    }
}
```

Option 3
---

Same as option 2 conceptually in terms of only making the "create a call invocation details" virtual, but using composition instead of inheritance:

```csharp
public sealed class MathClient
{
    private readonly AbstractMathCallProvider callProvider;

    public MathClient(AbstractMathCallProvider callProvider)
    {
        this.callProvider = callProvider;
    }

    public DivReply Div(DivArgs args, CallOptions options)
    {
        var call = callProvider.CreateDivCall(options);
        return Calls.BlockingUnaryCall(call, request);
    }

    public AsyncUnaryCall<DivReply> DivAsync(DivArgs args, CallOptions options)
    {
        var call = callProvider.CreateDivCall(options);
        return Calls.AsyncUnaryCall(call, request);
    }

    public DivReply Div(DivArgs args, Metadata headers = null, DateTime? deadline = null, CancellationToken cancellationToken = default(CancellationToken))
    {
        return Div(args, new CallOptions(headers, deadline, cancellationToken));
    }

    public AsyncUnaryCall<DivReply> DivAsync(DivArgs args, Metadata headers = null, DateTime? deadline = null, CancellationToken cancellationToken = default(CancellationToken))
    {
        return DivAsync(args, new CallOptions(headers, deadline, cancellationToken));
    }
}

public abstract class AbstractMathCallProvider
{
    public virtual CallInvocationDetails<DivArgs, DivReply> CreateDivCall(CallOptions options)
    {
        return SomeHelperType.CreateCall(channel, __Method_Div, options);
    }
}

public sealed class MathCallProvider : AbstractMathCallProvider
{
    private readonly Channel channel;

    // TODO: Header interceptors? Host? Combine them all into one concept?
    public MathClient(Channel channel)
    {
        this.channel = channel; // Plus validation :)
    }

    protected override CallInvocationDetails<DivArgs, DivReply> CreateDivCall(CallOptions options)
    {
        return SomeHelperType.CreateCall(channel, __Method_Div, options);
    }
}
```

Pros and cons:

- Option 1 has the fewest methods and is closest to what we have now, but in order to create a test double you need to distinguish between sync and async
- Option 2 has a single virtual method to be implemented per operation (which is logical IMO) but ends up feeling odd in terms of inheritance
- Option 3 feels logically clean, but we end up with more logical layers, which could cause confusion.

Server side
---

For the server side, I'd propose just an abstract class with virtual methods throwing `NotImplementedException` instead of the interface.

**Comment by jtattermusch** [comment]
@chrisdunelm   I think we can keep the server side interface -  that way we'll not break existing users. 
Also having the interface might come handy in some situations like server-side mocking.

**Comment by jtattermusch** [comment]
@jskeet   this issue is only for the server side, and most of the documentation you posted is about the client side. (The reason why I created an issue for the server side only so far is that it seems that on the server side we have more clarity and agreement about what it would take to fix the problem).

Mind creating a separate issue for the client side  and moving the post there?

**Comment by jskeet** [comment]
Yup - will do, but it'll probably be tomorrow :)

**Comment by jtattermusch** [comment]
@jskeet  thanks!

**Comment by chrisdunelm** [comment]
Keeping the server-side interface means that all RPC method additions break implementations of the interface. Even if we provide the abstract class, some users will always directly implement the interface.
So for versioning, this means that we either:
* Break semantic versioning, by only doing a minor version bump on RPC method addition
* Keep semantic versioning, by doing a major version bump on RPC method addition

How much pain would be caused by removing the interface now?

As I understand it, mocking (and DI) works fine with virtual methods in an abstract class; just as well as with an interface.

**Comment by jtattermusch** [comment]
Ad removing the interface:
I'd be generally in favor of it, but removing the interface would mean we would break all the existing users. We've committed to API stability with our beta release, so introducing such breaking change is not something we can do easily and I'd have to get approved by several members of our team.

**Comment by jskeet** [comment]
Hmm. I would still at least ask for that approval - it feels like it's a matter of breaking a few people now vs potentially a lot of people later on. (Allegedly, `make` only uses tabs because by the time it had been noted as a pain point, there were "dozens" of programs that would need fixing...)

**Comment by chrisdunelm** [comment]
Possibly also relevant: #5498 "Consider removing the flattened CallOptions overload from C# GRPC generated client service code?"

**Comment by jtattermusch** [comment]
@jskeet   agreed.

**Comment by jtattermusch** [comment]
Fixed by #5928
